pkg/storage/filepath/jsonfile_rest.go

```diff
@@ -206,110 +201,125 @@ func (f *filepathREST) Update(
 	forceAllowCreate bool,
 	options *metav1.UpdateOptions,
 ) (runtime.Object, bool, error) {
-	isCreate := false
-	oldObj, err := f.Get(ctx, name, nil)
-	if err != nil {
-		if !forceAllowCreate {
-			return nil, false, err
-		}
-		isCreate = true
-	}
-	oldVersion, err := resourceVersion(oldObj)
-	if err != nil {
-		return nil, false, err
-	}
-
-	// TODO: should not be necessary, verify Get works before creating filepath
-	if f.NamespaceScoped() {
-		// ensures namespace dir
-		ns, ok := genericapirequest.NamespaceFrom(ctx)
-		if !ok {
-			return nil, false, ErrNamespaceNotExists
+	var isCreate bool
+	var isDelete bool
+	obj, err := f.guaranteedUpdate(ctx, name, func(input runtime.Object) (output runtime.Object, err error) {
```

Comment:
The name here comes from K8s/etcd storage code, I can add some comments (and I'm open to suggestions on a different name!)

There are two `UpdatedObjectInfo` variants:
* Replace (`PUT`) -> ignore `storageObj`, just return `newObj`
* Patch -> take `storageObj`, apply patch

Remember that patch is a partial object replace, i.e. specific fields, with last-one-wins behavior.

Thus, by retrying a patch on a _storage-level_ concurrency failure (a write went in since we fetched the obj to mutate but before we got to write it), we'll keep fetching the new version, applying our mutations, and attempting persistence as long as we get a version conflict _from the storage layer_ (`MemoryFS::Write`). We'll never fail the _handler layer_ (`filepathREST::Update`) version check because we've copied+mutated the storage object each time, so we're always "up-to-date" from the handler's perspective.

It's totally possible to hit this same scenario for a replace (obj was current at start of update but out of date by time we wrote it to storage)! But on the retry, the updated `storageObj` version will now no longer match our `newObj` version, so the _handler layer_ (`filepathREST::Update`) will return a conflict error at that point.